### PR TITLE
Replace tinygltf with a safer homegrown loader

### PR DESCRIPTION
**Pull request by appgurueu**
In the process of this I uncovered some questionable loader behavior which I've annotated with `FIXME`s. Tests are passing.

### CODE REVIEW

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 1]
```diff
@@ -63,7 +62,7 @@ unsigned char CGLTFMeshFileLoader::BufferOffset::at(
 	return m_buf.at(m_offset + fromOffset);
 }
 
-CGLTFMeshFileLoader::CGLTFMeshFileLoader() noexcept
+CGLTFMeshFileLoader::CGLTFMeshFileLoader()
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp]: Why is `noexcept` removed?

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 2]
```diff
@@ -81,13 +80,24 @@ bool CGLTFMeshFileLoader::isALoadableFileExtension(
 */
 IAnimatedMesh* CGLTFMeshFileLoader::createMesh(io::IReadFile* file)
 {
-	tinygltf::Model model {};
+	if (file->getSize() <= 0) {
+		return nullptr;
+	}
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp:85]: Shouldn't it be `==`? I think this also needs a unit test.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 3]
```diff
@@ -81,13 +80,24 @@ bool CGLTFMeshFileLoader::isALoadableFileExtension(
...
+		return nullptr;
+	}
 
-	if (file->getSize() <= 0 || !tryParseGLTF(file, model)) {
+	if (!(model->buffers.has_value()
+			&& model->bufferViews.has_value()
+			&& model->accessors.has_value()
+			&& model->meshes.has_value()
+			&& model->nodes.has_value())) {
+		throw std::runtime_error("what");
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp]: sfan5 will be overjoyed to see you remembered not to leave the error message empty.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 4]
```diff
@@ -123,13 +133,7 @@ void CGLTFMeshFileLoader::loadPrimitives(
 }
 
 CGLTFMeshFileLoader::MeshExtractor::MeshExtractor(
-		const tinygltf::Model& model) noexcept
-	: m_model(model)
-{
-}
-
-CGLTFMeshFileLoader::MeshExtractor::MeshExtractor(
-		const tinygltf::Model&& model) noexcept
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp]: Rule of 5. If one of the destructors, copy or move destructors is manually defined, then you probably have to define all of them.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 5]
```diff
@@ -170,14 +177,14 @@ std::vector<video::S3DVertex> CGLTFMeshFileLoader::MeshExtractor::getVertices(
 
 	const auto normalAccessorIdx = getNormalAccessorIdx(
 			meshIdx, primitiveIdx);
-	if (normalAccessorIdx != static_cast<std::size_t>(-1)) {
-		copyNormals(normalAccessorIdx, vertices);
+	if (normalAccessorIdx.has_value()) {
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp:185]: Sweet! This is much cleaner now!

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 6]
```diff
@@ -412,70 +428,53 @@ std::size_t CGLTFMeshFileLoader::MeshExtractor::getPositionAccessorIdx(
...
-		return -1;
-	} else {
-		return result->second;
-	}
+	return m_model.meshes->at(meshIdx).primitives[primitiveIdx].attributes.normal;
 }
 
 /**
- * The index of the accessor that contains the NORMALs.
+ * The index of the accessor that contains the TEXCOORDs.
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp:444]: Good catch.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp [hunk 7]
```diff
@@ -412,70 +428,53 @@ std::size_t CGLTFMeshFileLoader::MeshExtractor::getPositionAccessorIdx(
...
-        }
-
-	if (err != "") {
-                os::Printer::log(err.c_str(), ELL_ERROR);
-		return false;
+	auto size = file->getSize();
+	auto buf = std::make_unique<char[]>(size + 1);
+	file->read(buf.get(), size);
+	// We probably don't need this, but add it just to be sure.
+	buf[size] = '\0';
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.cpp:468]: Yes, this is a buffer, not a null-terminated string. Don't add a null-terminator. What we should do is use a span to keep the buffer size together with the buffer.

File: source/Irrlicht/CGLTFMeshFileLoader.h [hunk 8]
```diff
@@ -25,7 +25,7 @@ namespace scene
 class CGLTFMeshFileLoader : public IMeshLoader
 {
 public:
-	CGLTFMeshFileLoader() noexcept;
+	CGLTFMeshFileLoader();
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.h]: Again, why is `noexcept` removed?

File: source/Irrlicht/CGLTFMeshFileLoader.h [hunk 9]
```diff
@@ -52,9 +52,7 @@ class CGLTFMeshFileLoader : public IMeshLoader
 	public:
 		using vertex_t = video::S3DVertex;
 
-		MeshExtractor(const tinygltf::Model& model) noexcept;
-
-		MeshExtractor(const tinygltf::Model&& model) noexcept;
+		MeshExtractor(const tiniergltf::GlTF& model);
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.h]: Why is `noexcept` removed?

File: source/Irrlicht/CGLTFMeshFileLoader.h [hunk 10]
```diff
@@ -52,9 +52,7 @@ class CGLTFMeshFileLoader : public IMeshLoader
 	public:
 		using vertex_t = video::S3DVertex;
 
-		MeshExtractor(const tinygltf::Model& model) noexcept;
-
-		MeshExtractor(const tinygltf::Model&& model) noexcept;
```
Comment (JosiahWI) [inline source/Irrlicht/CGLTFMeshFileLoader.h:57]: Rule of 5. Keep the move constructor.

File: source/Irrlicht/CMakeLists.txt [hunk 11]
```diff
@@ -331,7 +331,7 @@ set(link_libs
 	"${ZLIB_LIBRARY}"
 	"${JPEG_LIBRARY}"
 	"${PNG_LIBRARY}"
-	tinygltf::tinygltf
+	tiniergltf
```
Comment (JosiahWI) [inline source/Irrlicht/CMakeLists.txt]: Target should be namespaced.

File: source/Irrlicht/CMakeLists.txt [hunk 12]
```diff
@@ -292,21 +292,21 @@ endif()
...
-	foreach(TINYGLTF_DEF ${tinygltf_COMPILE_DEFS})
-		add_definitions("-D${TINYGLTF_DEF}")
-	endforeach(TINYGLTF_DEF)
+	foreach(TINIERGLTF_DEF ${tiniergltf_COMPILE_DEFS})
+		add_definitions("-D${TINIERGLTF_DEF}")
+	endforeach(TINIERGLTF_DEF)
 else()
-	add_compile_definitions(${tinygltf_COMPILE_DEFS})
+	add_compile_definitions(${tiniergltf_COMPILE_DEFS})
 endif()
```
Comment (JosiahWI) [inline source/Irrlicht/CMakeLists.txt]: This is no longer needed.

**Review by JosiahWI — CHANGES_REQUESTED** [review]
The tiniergltf library should be PRd separately.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp (hunk 7 above)
Comment (appgurueu) [reply source/Irrlicht/CGLTFMeshFileLoader.cpp:468]: I based this on the https://github.com/open-source-parsers/jsoncpp/blob/master/example/readFromString/readFromString.cpp example, so I'm not sure whether jsoncpp requires it to be null-terminated. I would assume not, because it requires you to pass the end pointer.

The [docs](http://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_char_reader.html#a65517004c4b5b1dc659ab966b3cea4cf) say "[t]he document must be a UTF-8 encoded string containing the document to read", so that doesn't really help :/

File: source/Irrlicht/CGLTFMeshFileLoader.cpp (hunk 3 above)
Comment (appgurueu) [reply source/Irrlicht/CGLTFMeshFileLoader.cpp]: oops, I forgot to remove this :sweat_smile:

File: source/Irrlicht/CGLTFMeshFileLoader.cpp (hunk 2 above)
Comment (appgurueu) [reply source/Irrlicht/CGLTFMeshFileLoader.cpp:85]: I basically just preserved the current logic, but this is juicy indeed... firstly, Irrlicht's ReadFile uses a `long` for this, so negative values are in principle possible.

(Even worse, if `file` is invalid, it seems the size field won't be initialized? :scream:)

Otherwise, this field is initialized using a call to `ftell`. The man page for `ftell` says:

> [...] Upon successful completion, [...] ftell() returns the current offset.  Otherwise, -1 is returned and errno is set to indicate the error.

So this can indeed be `-1` in case of an error.

We should probably also default-initialize size to `-1` in `CReadFile`.

> I think this also needs a unit test.

Don't we have `empty.gltf` for `== 0` already? (That said, an empty string also shouldn't get past `jsoncpp`, since it isn't valid JSON.)

Comment (JosiahWI) [reply source/Irrlicht/CGLTFMeshFileLoader.cpp:85]: Ah, so that's how it passed before, then. I think it'd be good to have a comment explaining why it's `<=`.

File: source/Irrlicht/CGLTFMeshFileLoader.cpp (hunk 7 above)
Comment (JosiahWI) [reply source/Irrlicht/CGLTFMeshFileLoader.cpp:468]: The docs don't say it has to be null-terminated, so I don't think it has to be, but since I've had time to think about it more, I'm fine with leaving this in.

File: source/Irrlicht/CReadFile.cpp [hunk 13]
```diff
@@ -3,6 +3,7 @@
 // For conditions of distribution and use, see copyright notice in irrlicht.h
 
 #include "CReadFile.h"
+#include <istream>
```
Comment (JosiahWI) [inline source/Irrlicht/CReadFile.cpp]: What's the reason for this change?

Comment (appgurueu) [reply source/Irrlicht/CReadFile.cpp]: oops
